x-pack/plugins/security_solution/server/endpoint/routes/metadata/handlers.ts

```diff
@@ -161,6 +167,83 @@ export const getMetadataListRequestHandler = function (
   };
 };
 
+export function getMetadataListRequestHandlerV2(
+  endpointAppContext: EndpointAppContext,
+  logger: Logger
+): RequestHandler<
+  unknown,
+  TypeOf<typeof GetMetadataListRequestSchemaV2.query>,
+  unknown,
+  SecuritySolutionRequestHandlerContext
+> {
+  return async (context, request, response) => {
+    const endpointMetadataService = endpointAppContext.service.getEndpointMetadataService();
+    if (!endpointMetadataService) {
+      throw new EndpointError('endpoint metadata service not available');
+    }
+
+    let doesUnitedIndexExist = false;
+    let didUnitedIndexError = false;
+    let body: MetadataListResponse = {
+      data: [],
+      total: 0,
+      page: 0,
+      pageSize: 0,
+    };
+
+    try {
+      doesUnitedIndexExist = await endpointMetadataService.doesUnitedIndexExist(
+        context.core.elasticsearch.client.asCurrentUser
+      );
+    } catch (error) {
+      // for better UX, try legacy query instead of immediately failing on united index error
+      didUnitedIndexError = true;
+    }
+
+    // If no unified Index present, then perform a search using the legacy approach
+    if (!doesUnitedIndexExist || didUnitedIndexError) {
+      const endpointPolicies = await getAllEndpointPackagePolicies(
+        endpointAppContext.service.getPackagePolicyService(),
+        context.core.savedObjects.client
+      );
+
+      const legacyResponse = await legacyListMetadataQuery(
+        context,
+        endpointAppContext,
+        logger,
+        endpointPolicies,
+        request.query
+      );
+      body = {
+        data: legacyResponse.hosts,
+        total: legacyResponse.total,
+        page: request.query.page,
+        pageSize: request.query.pageSize,
+      };
+      return response.ok({ body });
+    }
+
+    // Unified index is installed and being used - perform search using new approach
+    try {
+      const { data, total } = await endpointMetadataService.getHostMetadataList(
+        context.core.elasticsearch.client.asCurrentUser,
+        request.query
+      );
+
+      body = {
+        data,
+        total,
+        page: request.query.page,
```

Comment:
I changed to this because I removed `page` and `pageSize` from `getHostMetadataList` since it just returns those directly from the received params anyways.